snakemake/deployment/spack.py

```diff
@@ -0,0 +1,232 @@
+        else:
+            env_file = env_file.get_path_or_uri()
+
+        env_path = self.path
+
+        # Check for broken environment
+        self.check_broken_environment(env_path, dryrun)
+
+        # Create environment if not already present.
+        if not os.path.exists(env_path):
+            if dryrun:
+                logger.info(
+                    "Spack environment {} will be created.".format(
+                        self.file.simplify_path()
+                    )
+                )
+                return env_path
+
+            logger.info(
+                "Creating spack environment {}...".format(self.file.simplify_path())
+            )
+
+            try:
+                # Touch "start" flag file
+                os.makedirs(env_path, exist_ok=True)
+                Path(os.path.join(env_path, "env_setup_start")).touch()
+
+                # Copy env file to env_path so we can see what an
+                # environment in .snakemake/spack contains.
+                target_env_file = os.path.join(env_path, "spack.yaml")
+                shutil.copyfile(env_file, target_env_file)
+
+                logger.info("Downloading and installing remote packages.")
+                create = " ".join(
+                    [
+                        "spack",
+                        "env",
+                        "create",
+                        "--dir '{}'".format(env_path),
+                        target_env_file,
+                    ]
+                )
+
+                ## Finally, source and install (also concretized), and deactivate
+                install = "spack install"
+                activate = "eval `spack env activate --sh %s`; %s" % (env_path, install)
+
+                for cmd in [create, activate]:
+                    logger.info(cmd)
+                    out = shell.check_output(
+                        cmd, stderr=subprocess.STDOUT, universal_newlines=True
+                    )
+
+                # Touch "done" flag file
+                Path(os.path.join(env_path, "env_setup_done")).touch()
+                logger.debug(out)
+                logger.info(
+                    "Environment for {} created (location: {})".format(
+                        os.path.relpath(env_file), os.path.relpath(env_path)
+                    )
+                )
+            except subprocess.CalledProcessError as e:
+                # remove potential partially installed environment
+                shutil.rmtree(env_path, ignore_errors=True)
+                raise CreateSpackEnvironmentException(
+                    "Could not create spack environment from {}:\n".format(env_file)
+                    + e.output
+                )
+
+        if tmp_file:
+            # temporary file was created
+            os.remove(tmp_file)
+
+        return env_path
+
+    def __eq__(self, other):
+        if isinstance(other, Env):
+            return self.file == other.file
+        return False
+
+
+class Spack:
```

Comment:
I think this class can also share a base class with the conda implementation.

Reply:
oh definitely - I'm seeing a lot of duplication that (assuming I give the ability for the base to have a container and then spack doesn't support that) can simplify the code quite a bit - working now will update soon!